src/hotspot/share/memory/metaspace.cpp

```diff
@@ -995,23 +995,30 @@ const char* Metaspace::metadata_type_name(Metaspace::MetadataType mdtype) {
   }
 }
 
-void Metaspace::purge() {
+void Metaspace::purge(bool classes_unloaded) {
```

Comment:
Would it not have been more logical to heave both `MetaspaceCritical_lock` and `MetaspaceCriticalAllocation::process();` out of this function? As it is, `Metaspace::purge`, called with false, is now a no-op that only does things on behalf of the GC. It has nothing to do with the functioning of the allocator.

Reply:
I understand what you mean. On the other hand, if you take a look at Metaspace::allocate, it does quite a lot more than being an allocator, for example, it calls into the GC code (which performs the other side of the metaspace critical allocations).

I wouldn't be opposed if anyone wanted to restructure this.